Added Timebomb functionality to Features

AidanLaycock opened the pull request:
This PR added the ability to set a feature to expire during creation. It then throws an exception, designed to break CI so that technical debt can be removed.

Users can also create long living Features when using Timebombs by leaving an expiry as null (This could be useful for Beta Features).

ksassnowski commented:
I feel like this is rather dangerous in its current state as it could make the production system randomly explode unless you remembered to deploy a new version in time. Ideally, this should only happen during CI (so probably depending on `APP_ENV`). What do you think?

ksassnowski commented on tests/Unit/FeatureTimeBombTest.php:
```diff
@@ -0,0 +1,87 @@
+
+    expect(Feature::all())->toHaveCount(1);
+    // Assert No Exception is thrown
+    $this->assertTrue(true);
```
I think this is redundant since you already assert that the feature was created. If there had been an exception, the test would have failed before you got to this point in the test.

AidanLaycock replied:
Thank you, I've removed this test now.

ksassnowski commented:
This might come down to a stylistic choice but I'm not a big fan of the dummy `$this->assertTrue(true)` assertions to assert that no exception was thrown. The reason is that you're not _really_ asserting that no exception was thrown, you're asserting that `true === true` and sort of implicitly rely on the fact that in order to even reach this code, no exception could have been thrown.

My suggestion would be to replace all the `$this->assertTrue(true)` calls with the following:

```php
it('should not throw an exception', function () {
    $exception = null;

    try {
        $foo->methodThatMightThrowException();
    catch (Throwable $exception) {
        // If an exception was thrown, it will be assigned to `$exception`.
        //
        // To make it more explicit what's happenning, you could alternatively 
        // use two different variables for the exception and explicitly assign 
        // the caught exception to the outer variable.
    }

    // Explicitly test that no exception was caught.
    assertNull($e, 'Unexpected assertion was thrown');
});
```

While this is rather verbose, it very explicitly asserts against the thing the test is claiming to test: that there was no exception.

Some food for thought :)

AidanLaycock commented:
Hey @ksassnowski - Thanks for the feedback, I'll have a look at those later today and update the PR :)

ksassnowski commented:
Just FYI, I'm not a maintainer of this project so these are just suggestions. @JustSteveKing has the final say :)

JustSteveKing commented:
I think @ksassnowski has brought up some good points worth looking at. The last thing I want to do is to introduce changes where production systems suddenly start exploding because of the time bomb feature.

Good catch on the tests too!

AidanLaycock commented:
@ksassnowski - Thank you again for the feedback :) I've implemented those points.

@JustSteveKing - This should hopefully be ready to be checked/merged now, let me know if you want anything else adding, changing! :)

ksassnowski commented on config/feature-flags.php:
```diff
@@ -36,4 +36,28 @@
+     | the array as empty.
+     |
+     */
+     'time_bomb_environments' => ['production']
```
I find the name `time_bomb_environments` a little counter-intuitive. Without reading the comment, I would have assumed that these are the environments in which the time bombs are _active_, not the other way around. 

Another thing is that my first instinct have would be to have the time bombs be deactivated in every environment and to explicitly activate them for certain environments. I think this would reduce the chance of accidentally activating it in production because of a typo or something similar.

JustSteveKing replied:
I have to agree here, time bomb feature should be enabled per environment and off by default - not something you would typically have in staging/qa

ksassnowski commented on src/Models/Feature.php:
```diff
@@ -4,25 +4,43 @@
+    public static function booted()
+    {
+        static::retrieved(function(Feature $feature) {
+            if(
```
Again, more of a stylistic thing, but I find this section a bit difficult to read with all the nested conditionals and ternary expressions. I think this would benefit from a couple of intermediate variables.

JustSteveKing replied:
I'm all for clean code - but nested conditionals make this harder to read and test, which means it could cause issues further down the line

AidanLaycock replied:
Thank you both, I've refactored this now :)

ksassnowski commented on src/Models/Feature.php:
```diff
@@ -4,25 +4,43 @@
+                config('feature-flags.enable_time_bombs')
+                && ! App::environment(config('feature-flags.time_bomb_environments'))
+            ) {
+                return (! is_null($feature->expires_at) && Carbon::now() >= $feature->expires_at) ? throw new Exception(sprintf('The Feature has expired - %s', $feature->name)) : true;
```
This ternary expression is a little strange since the `if` branch essentially evaluates to

```php
return throw new Exception(...);
```

I think expanding this out to an actual if-statement would make this much more readable. The condition can also be further simplified by using `Carbon::isAfter(...)`. This would allow us to get rid of the null check since `Carbon::isAfter(null)` evaluates to `false`.

```php
$featureHasExpired = Carbon::now()->isAfter($feature->expired_at);

if ($featureHasExpired) {
    throw new Exception(sprintf('The Feature has expired - %s', $feature->name));
}

return true;
```

AidanLaycock replied:
Cool, I've updated this :) I think it's because I'm used to using ternary ifs in JS more often :)

ksassnowski commented on tests/Unit/FeatureTimeBombTest.php:
```diff
@@ -24,19 +36,6 @@
        ->toHaveCount(1);
 });
 
-it('Does not throw an Exception when time bombs are disabled', function(): void {
```
Sorry, my last comment might have been misleading.  I meant that I found the `$this->assertTrue(true)` assertion unnecessary, not the entire test. I think this test should definitely be kept but maybe use the `tryException` helper you added like in the other tests

AidanLaycock replied:
Ah sorry I get you now :) I've readded this test back in.

AidanLaycock commented:
Hey @JustSteveKing - I was wondering if you'd had a chance to have another look at this, and whether it could be merged?

JustSteveKing commented:
Hey @AidanLaycock 

Sorry my GitHub notifications seem to be having a bit of a funny time!

So ideally I would like to see the time bomb feature as opt in, based on an ENV variable, with a default set to disable them. I love this idea, but introducing it when it's already published means it needs to mirror what's there without breaking where possible

JustSteveKing commented:
Go for it @AidanLaycock - I will keep my eye on PRs tomorrow 💪

AidanLaycock commented:
Hey @JustSteveKing - That's great, this should hopefully be ready for merging now! I've added more info in the ReadMe about Timebombs. And I've also updated a few return types that I spotted whilst reopening the repo.

Do let me know if you want anything else changing! And please feel free to rework the readMe as you see fit!

JustSteveKing commented:
Merged and released! Thanks @AidanLaycock
